utils/optimin/src/OptiMin.cpp

```diff
@@ -0,0 +1,443 @@
+// `afl-showmap.c`
+static constexpr uint32_t MAX_EDGE_FREQ = 8;
+
+static sys::TimePoint<>     StartTime, EndTime;
+static std::chrono::seconds Duration;
+
+static std::string AFLShowmapPath;
+static bool        TargetArgsHasAtAt = false;
+
+static cl::opt<std::string>  CorpusDir("i", cl::desc("Input directory"),
+                                      cl::value_desc("dir"), cl::Required);
+static cl::opt<std::string>  OutputDir("o", cl::desc("Output directory"),
+                                      cl::value_desc("dir"), cl::Required);
+static cl::opt<bool>         EdgesOnly("f", cl::desc("Include edge hit counts"),
+                               cl::init(true));
+static cl::opt<bool>         ShowProgBar("p", cl::desc("Display progress bar"));
+static cl::opt<std::string>  WeightsFile("w", cl::desc("Weights file"),
+                                        cl::value_desc("csv"));
+static cl::opt<std::string>  TargetProg(cl::Positional,
+                                       cl::desc("<target program>"),
+                                       cl::Required);
+static cl::list<std::string> TargetArgs(cl::ConsumeAfter,
+                                        cl::desc("[target args...]"));
+static cl::opt<std::string>  MemLimit(
+    "m", cl::desc("Memory limit for child process (default=none)"),
+    cl::value_desc("megs"), cl::init("none"));
+static cl::opt<std::string> Timeout(
+    "t", cl::desc("Run time limit for child process (default=none)"),
+    cl::value_desc("msec"), cl::init("none"));
+}  // anonymous namespace
+
+// -------------------------------------------------------------------------- //
+// Helper functions
+// -------------------------------------------------------------------------- //
+
+static void GetWeights(const MemoryBuffer &MB, WeightsMap &Weights) {
+  SmallVector<StringRef, 0> Lines;
+  MB.getBuffer().split(Lines, '\n');
+
+  unsigned Weight = 0;
+
+  for (const auto &Line : Lines) {
+    const auto &[Seed, WeightStr] = Line.split(',');
+
+    to_integer(WeightStr, Weight, 10);
+  }
+}
+
+[[nodiscard]] static std::error_code getAFLCoverage(const StringRef    Seed,
+                                                    AFLCoverageVector &Cov) {
+  Optional<StringRef> Redirects[] = {None, None, None};
+  std::error_code     EC;
+
+  // Create temporary output file
+  SmallString<64> OutputPath;
+  if (EC = sys::fs::createTemporaryFile("showmap", "txt", OutputPath))
+    return EC;
+
+  // Prepare afl-showmap arguments
+  SmallVector<StringRef, 12> AFLShowmapArgs{
+      AFLShowmapPath, "-m", MemLimit, "-t", Timeout, "-q", "-o", OutputPath};
+
+  if (TargetArgsHasAtAt)
+    AFLShowmapArgs.append({"-A", Seed});
+  else
+    Redirects[/* stdin */ 0] = Seed;
+
+  AFLShowmapArgs.append({"--", TargetProg});
+  AFLShowmapArgs.append(TargetArgs.begin(), TargetArgs.end());
+
+  // Run afl-showmap
+  const int RC = sys::ExecuteAndWait(AFLShowmapPath, AFLShowmapArgs,
```

Comment:
Here problably this should be replaced with the code in forkserver.h

Reply:
I did consider this, but given that I use the LLVM API so much already, I figured I may as well use it exclusively. Let me know if you want me to change this, however!